jmclient/jmclient/wallet_service.py

```diff
@@ -389,22 +399,26 @@ def transaction_monitor(self):
             if len(added_utxos) > 0 or len(removed_utxos) > 0 \
                     or txid in self.active_txs:
                 if confs == 0:
-                    for k in possible_keys:
-                        if k in self.callbacks["unconfirmed"]:
-                            for f in self.callbacks["unconfirmed"][k]:
-                                # True implies success, implies removal:
-                                if f(txd, txid):
-                                    self.callbacks["unconfirmed"][k].remove(f)
-                                    # keep monitoring for conf > 0:
-                                    self.active_txs[txid] = txd
```

Comment:
@AdamISZ: There were two bugs here:

* `f` is removed from `self.callbacks["unconfirmed"][k]` while iterating over that list. This is a no-no. Suppose `f` is at index 0 in the list. Removing `f` from the list will shift all remaining elements in the list down by one position, so the element at index 1 will move to index 0. Then the next iteration of the loop will examine the element at index 1. Oops! The element that moved into index 0 is never examined.
* The transaction is only added to the `self.active_txs` dict (previously, the `self.active_txids` list) if at least one `unconfirmed` callback returned `True`. Suppose that a particular TxID had only `confirmed` callbacks registered for it but no `unconfirmed` callbacks. Once the transaction is observed in the Bitcoin Core wallet, it will only be polled for confirmation if it is present in `self.active_txs`, but it won't be unless at least one `unconfirmed` callback returned `True` for that transaction.

Reply:
Re: update-in-loop, thanks for that - my memory is pretty weak on this, I have a vague sense that that issue was discussed, and tbh I thought it had been fixed/addressed. Oops, indeed :) I am a little confused as to why this has not caused more problems, presumably it is mainly because we don't usually have lists of more than one item.

Second point: yes, good catch there - I had it coded to so to speak 'reject' a notification if the unconfirmed callback returned False, but that makes no sense as there can be multiple independent callbacks registered. So we still keep the behaviour that a `False` return from a callback means it isn't removed. All good.